---
title: "Phylogenetic Tree Construction"
author: "Marian Schmidt"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Goals of this file 

The goal is to create a phylogenetic tree! 

1. Load in preprocessed phyloseq object.  
2. Create ASV fasta file from the phyloseq object. 
3. Align the 16S sequences from fasta file with MAFFT. 
4. Create a tree with FastTree2.  


## Load Packages & Phyloseq Object
```{r load-environment}
# phytools, ggtree, RColorBrewer
pacman::p_load(phytools, ggtree, RColorBrewer, install = FALSE)

# Load physeq 
load("data/02_PreProcessing/raw_preprocessed_physeq.RData")
raw_preprocessed_physeq
```

# Create Fasta File of ASV and their Sequences

This fasta file will be used to create our alignment in MAFFT

```{r create-preprocessed-fasta}
# pull out ASV seqs and ASV names 
asv_seq_df <- 
  raw_preprocessed_physeq@tax_table %>%
  data.frame() %>%
  dplyr::select(ASV, ASVseq)

#View(asv_seq_df)

# Add the > to make fasta header 
asv_seq_df$ASV <- paste0(">",asv_seq_df$ASV)
#View(asv_seq_df)

# Create the fasta object 
asv_seq_fasta <- c(rbind(asv_seq_df$ASV, asv_seq_df$ASVseq))
head(asv_seq_fasta)

# Write to a file 
write(asv_seq_fasta, 
      file = "data/03_Phylogenetic_Tree/preprocessed_ASVs.fasta")
```

# Align the 16S sequences from fasta file with MAFFT

Mar - you ended class here. :) 
# run bash code inside of our R markdown document to access MAFFT-- there are a few ways to do this-- you could just start the chunk name with bash, or do engine=.  We're using the export path because we're running the script in R (we went to the bioHPC user guide for mafft to get the code to copy and paste)

```{r run-mafft, engine = 'bash', engine.opts= '-l'}
#Write bash code to run MAFFT

export PATH=/programs/mafft/bin:$PATH

#change difrectories to provide the fasta file we made above
cd data/03_Phylogenetic_Tree/
pwd
# Set a seed-- could use the same seed as before for consistency
RANDOM=19

# Run Mafft
# For Now, we will use default options, could read-up on different parameters/options you run for your own project.  Since we're using default options, we need to note the version on our computer
#MAFFT automatically knows that it's a nucleotide alignment. #The carrot symbol in bash means you're writing a file to a place 
/programs/mafft/bin/mafft --auto preprocessed_ASVs.fasta > MAFFT_aligned_asvs.fasta

#change back to your directory

cd ../../
pwd ..
#If you wanted to know more about mafft, google mafft allignment- tips, viewy, how they suggest running mafft, etc.
#You can test your code in the terminal-For macs you can push command opt enter and it will send the code directly to the terminal.

```

# Now we need to navigate our input into fastree (we could google FASTREE2 phylogeny) google will tell us about microbesonline.org has good info about it and look at papers that have used it, we could maybe look at their code. 

#FastTree2 
# We went to the BioHPC site and found fastree in the user guide to find out how to use fastree do the export path because we're using it in a script
# the engine.opts= "-l" makes it so that when we're working in r, it is like working in our bash terminal--like any programs we have loaded, etc. will be there.

```{r run-fasttree, engine='bash', engine.opts= "-l"} 
#Provide export path to fasttree
export PATH=/programs/FastTree-2.1.11:$PATH

#cd into the alignment file folder
cd data/03_Phylogenetic_Tree/
pwd
FastTree -h
#Run Fasttree to generate phylogenetic tree
# Some of the paratmeters were are going to use-- nt (means nucleotide alignment, -gtr is a generalized time reversible model-- a nucleotide substitution model, which provides like 9 different parameters for different types of substitution rates in our phylogenetic tree over time.  We need to be mindful of what underlying assumptions are part of our model.) -fastest will speed up the model and reduce memory usage (that's recommended for datasets with greater than 50,000 sequences), we'll also add -log to make a log file because it's helpful to have-- you can decide if you want to have the log file instead of the html, then we'll have our input aligment file and need to specify the output tree file.

# because we exported it to our path, we can just use the Fasttree command
FastTree -nt -gtr -fastest -log FastTree.log MAFFT_aligned_asvs.fasta > ASVs_unrooted.tree

#Change back to the main directory
cd ../../
pwd
#or we could say 
echo "The working directory is $PWD"

```

#It's a good idea to look at all your files after you generate them
# Now we're going to use 

```{r session-info}
devtools::session_info()
```

