---
title: "Infer ASVs with DADA2"
author: "Julie Cardon"
date: "2024-02-21"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.align = "center", 
                      fig.path = "../figures/01_DADA2") #send any figure output to this folder
#only for the first chunk put ../ to get back to salinity gradient folder, 
```

# Goals of This File

 1. Use raw fastq files and generate quality plots to assess quality of reads.
 2. Filter and trim out bad sequenes and bases from our sequencing files.
 3. Write out fastq files with high quality sequences.
 4. We will evaluate the quality from our filter and trim.
 
# Load Libraries

```{r load-libraries}
#install.packages("devtools")
library(devtools)
#devtools::install_github("benjjneb/dada2")
library(dada2)
```

# Load Data

```{r load-data} 
# Set the raw fastq path to the raw seqencing files
# Path to the fastq files
raw_fastqs_path <- "data/01_DADA2/01_raw_gzipped_fastqs"
raw_fastqs_path

#What files are in this path? Intuition Check
list.files(raw_fastqs_path)

#How many files are there?
str(list.files(raw_fastqs_path))

# Create a variable to work with forward reads, and a variable to work with reverse reads

# Create a vector of forward reads
forward_reads <- list.files(raw_fastqs_path, pattern = "R1_001.fastq.gz", full.names = TRUE)
forward_reads
#Intuition Check
head(forward_reads)


# Create a vector of reverse reads
reverse_reads <- list.files(raw_fastqs_path, pattern = "R2_001.fastq.gz", full.names = TRUE)
reverse_reads
head(reverse_reads)

```

#Quality Plots
```{r raw-quality-plot}
# Randomly select 2 samples from dataset to evaluate
random_samples <- sample(1:length(reverse_reads), size = 2)
random_samples

#Calculate and plot qulaity of these two samples
plotQualityProfile(forward_reads[random_samples])
plotQualityProfile(reverse_reads[random_samples])

```

# Filter and Trim reads

#Prepare a placeholder for filtered reads
```{r prep-filtered-sequences} 
# vector of our samples, extract sample name from files
samples <- sapply(strsplit(basename(forward_reads), "_"), `[`,1) 
#Instuition Check
head(samples)

#Place filtered reads into filtered fastqs path
filtered_fastqs_path <- "data/01_DADA2/02_filtered_fastqs"

# create 2 variables: filtere_F, filtered_R

filtered_forward_reads <- file.path(filtered_fastqs_path, paste0(samples, "_R1_filtered.fastq.gz"))

filtered_reverse_reads <- file.path(filtered_fastqs_path, paste0(samples, "_R2_filtered.fastq.gz")) 





```









